\

laboratory of computational & structural biology
------------------------------------------------

<div id="header">

  ----------------------------------- ------------------------------------------------------------------------ ---------------------------------------- ------------------------------------
  ![logo CSIC](./pics/logoCSIC.png)   [Estación Experimental Aula Dei](http://www.eead.csic.es/)\              [Fundación ARAID](http://www.araid.es)   ![logo ARID](./pics/logoARAID.gif)
                                      [Consejo Superior de Investigaciones Científicas](http://www.csic.es/)                                            
  ----------------------------------- ------------------------------------------------------------------------ ---------------------------------------- ------------------------------------

</div>

<div id="content">

-   [Home](index.html)
-   [Lab members](staff.html)
-   [Our collaborators](collaborators.html)
-   [Publications](publications_computational_biology_bioinformatics.html)
-   [Software](software_computational_biology_bioinformatics.html)
-   [Material
    educativo](material_didactico_biologia_computacional_bioinformatica.html)

<div id="colOne">

Our resources
[RSAT::Plants](http://plants.rsat.eu "A plant-dedicated server for the analysis of regulatory sequences")\
[BARLEYMAP](http://floresta.eead.csic.es/barleymap "tool to search the position of barley genetic markers on the Barley Physical Map and the POPSEQ map")\
[GET\_HOMOLOGUES](https://github.com/eead-csic-compbio/get_homologues "versatile package for pan-genome analysis")\
[footprintDB](http://floresta.eead.csic.es/footprintdb "database of transcription factors with annotated cis elements and binding interfaces, includes 3D-footprint")\
[3D-footprint](http://floresta.eead.csic.es/3dfootprint "database for the structural analysis of protein-DNA complexes")\
[*more*](./software_computational_biology_bioinformatics.html)
\
**blog:**
[\#!/perl/bioinfo](http://bioinfoperl.blogspot.com/ "bitácora de bioinformática y programación (in Spanish)")\
\
**postal address**\
Laboratorio de Biología Computacional y Estructural, Estación
Experimental Aula Dei-CSIC, Av. Montañana 1.005, 50059 Zaragoza (Spain)\
\
**phone** (+34) 976716089\
\
  --------------------------------------------- ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  ![public transport logo](./pics/public.png)   [*how to get here*](./HOWTO_get_to_Zaragoza.html)
  ![bicilogo](./pics/bici.png)                  *[rutas en bicicleta de Zaragoza al campus de Aula Dei](http://maps.google.es/maps/ms?ie=UTF8&hq=&hnear=Zaragoza,+Aragón&t=h&hl=es&msa=0&msid=103443040866049081278.0004716b3cad43164d2cf&ll=41.687912,-0.828438&spn=0.08653,0.222816&z=13)*
  --------------------------------------------- ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

</div>

<div id="colTwo">

<div id="welcome">

**Research topics/[objetivos del labo](javascript:void(0))**
<div id="lineas">

-   genomics and transcriptional regulation of grasses, in collaboration
    with [Ana Casas](http://www.eead.csic.es/home/staffinfo?Id=67),
    [Ernesto Igartua](http://www.eead.csic.es/home/staffinfo?Id=69),
    [Pilar Catalán](http://bifi.es/bioflora/research-team/profile1.html)
    and the [RSAT team](http://rsat.eead.csic.es/plants/people.php)
-   evolutionary and structural analysis of transcription factors and
    disordered proteins in plants
-   phylogenomic analyses and markers for breeding, molecular
    diagnostics and metagenomics, in collaboration with [Pablo
    Vinuesa (UNAM,México)](http://www.ccg.unam.mx/%7Evinuesa/)

</div>

<div class="navi">

</div>

[](){.prev}
<div class="scrollable">

<div id="thumbs">

<div>

+--------------------------------------+--------------------------------------+
| ![Maximum-likelihood pan-genome      | Maximum-likelihood pan-genome        |
| phylogeny](./thumbs/pangenometree.jp | phylogeny estimated with IQ-TREE     |
| g)                                   | from the consensus pan-genome        |
|                                      | clusters of *Stenotrophomonas*       |
|                                      | displayed in the Venn diagram        |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![Vrs1                               | a\) Alignment of Vrs1.b2, vrs1.a2, a |
| alleles](./thumbs/Vrs1MSA.gif)       | nd Vrs1.b5 DNA partial sequences of  |
|                                      | exon 2. b) Multiple alignment of pro |
|                                      | tein sequences of different alleles  |
|                                      | of the Vrs1 gene. c) Barley spikes f |
|                                      | rom Spanish landraces SBCC039        |
|                                      | (vrs1.a2) six-rowed (left) and SBCC1 |
|                                      | 55 (Vrs1.b5) two-rowed (right)       |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![pangenome of Brachypodium          | Gene-based pan-genome of             |
| distachyon](./thumbs/pangenomeBdis.j | *Brachypodium distachyon*. B,C)      |
| pg)                                  | Core, softcore, shell, cloud genes.  |
|                                      | D) % coverage of 37,886              |
|                                      | high-confidence pan-genes by short   |
|                                      | read data sets from 49 lines.        |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![Nuclear and plastid filogenomic    | Nuclear and plastid filogenomic      |
| trees of &lt;i&gt;Brachypodium       | trees of *Brachypodium distachyon*   |
| distachyon&lt;/i&gt;                 | ecotypes.                            |
| ecotypes](./thumbs/nuclear_cp_Bdis_t |                                      |
| rees.png)                            |                                      |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![pangenome                          | Pan-genome of CDS annotated in 19    |
| Athaliana](./thumbs/pangenomeAtha.jp | Arabidopsis thaliana ecotypes. A)    |
| g)                                   | Occupancy of CDS clusters. B)        |
|                                      | Pan-genome growth simulations. C)    |
|                                      | Mean expression of CDS in root,      |
|                                      | seedling and floral buds. D) dN/dS   |
|                                      | ratio of single-copy CDS clusters.   |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![MEBS                               | Schematic representation of the 4    |
| flowchart](./thumbs/MEBSflowchart.jp | stages of the MEBS algorithm         |
| g)                                   | focusing on the Sulphur-cycle.       |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![TF disordered residues vs          | Total disordered residues in         |
| tissues](./thumbs/disorderTF.png)    | transcription factors (PONDR VLS2b)  |
|                                      | vs number of tissues in several      |
|                                      | plant species.                       |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![barley drought                     | Design of stress treatments and leaf |
| experiment](./thumbs/barleydrought.j | water potential of barley SBCC073    |
| pg)                                  | (73) and Scarlett (SC) plants.       |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![heterozygous                       | Heterozygous mappings in barley      |
| mappings](./thumbs/heterozygous_mapp | occur when relevant loci are absent  |
| ings.jpeg)                           | from the reference genome assembly.  |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![nucleosome                         | Model of nucleosome sandwich         |
| GFP](./thumbs/nucleosomeGFP.jpg)     | incorporating a single H2BGFP        |
|                                      | molecule (in red). The steric        |
|                                      | difficulty to accommodate the GFP    |
|                                      | tag increases if we consider that    |
|                                      | each nucleosome may contain up to    |
|                                      | two molecules of GFP.                |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![interface annotation in            | Annotation of interface residues in  |
| footDB](./thumbs/footDB.gif)         | footprintDB.                         |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![interfaces of R2R3 MYB             | \(A) Sequence logos of the R2 and R3 |
| TFs](./thumbs/myb_interfaces.png)    |  recognition α-helices of Arabidopsi |
|                                      | s                                    |
|                                      | thaliana R2R3-MYB DBD subgroups (lef |
|                                      | t) and of the bound cis-elements     |
|                                      | assayed on the Y1H experiments, clas |
|                                      | sified into two groups (right).      |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![BARLEYMAP                          | BARLEYMAP pipeline                   |
| pipeline](./thumbs/barleymap.gif)    |                                      |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![plant                              | Distribution of predicted disordered |
| disorder](./thumbs/plant_disorder.jp | segments with length&gt;30 in        |
| g)                                   | plants. Disorder in chloroplast      |
|                                      | (Cp), mitochondria (Mt) and nuclear  |
|                                      | (Nu) proteomes are shown.            |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![footprint logo of restriction      | footprint of dimeric complex         |
| enzyme                               | [2bam\_AB](http://floresta.eead.csic |
| BamHI](./thumbs/2bam_AB_footprint.pn | .es/3dpwm/complexes/2bam_AB.html)\   |
| g)                                   | (enzyme BamHI)                       |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![active transcriptional subnetworks | A few active transcriptional         |
| in Escherichia                       | subnetworks in *Escherichia coli*    |
| coli](./thumbs/RNnets.jpg)           | (details                             |
|                                      | [here](http://nar.oxfordjournals.org |
|                                      | /cgi/reprint/gkq612v1))              |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![ML tree of plant-like FADS         | ML tree of plant-like FADS proteins  |
| proteins](./thumbs/fads_malign.png)  | (details                             |
|                                      | [here](http://www.biomedcentral.com/ |
|                                      | 1471-2148/10/311))                   |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![Multiple interface alignment of    | Multiple interface alignment of zinc |
| zinc fingers](./thumbs/ZFmalign.png) | fingers (details                     |
|                                      | [here](http://www3.interscience.wile |
|                                      | y.com/cgi-bin/fulltext/122477756/PDF |
|                                      | START))                              |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![flowchart of DNAPROT footprint     | DNAPROT algorithm (details           |
| algorithm](./thumbs/dnaprot.jpg)     | [here](http://www.biomedcentral.com/ |
|                                      | 1471-2105/9/436/figure/F2))          |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![3D-fooprint of a comparative model | structure-based position weight      |
| of Escherichia coli                  | matrix for *Escherichia coli* FNR,   |
| FNR](./thumbs/logos_models2008.png)  | modelled with TFmodeller\            |
|                                      | (read more                           |
|                                      | [here](http://www.biomedcentral.com/ |
|                                      | 1471-2105/9/436/figure/F5))          |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![TFmodeller homology model of a     | comparative model of a protein-DNA   |
| protein-DNA                          | complex\                             |
| complex](./thumbs/tfmodeller_model.g | (read more                           |
| if)                                  | [here](http://www.eead.csic.es/compb |
|                                      | io/papers/tfmodeller2007.pdf))       |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![evolutionary correlation between   | conceptual model of evolutionary     |
| binding specificity and regulatory   | role of binding specificity in       |
| role in                              | bacterial regulatory networks\       |
| Bacteria](./thumbs/specificity2008.p | (read more                           |
| ng)                                  | [here](http://www.eead.csic.es/compb |
|                                      | io/papers/prok_specificity.pdf))     |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![alpha parameter and codon usage    | genomic benchmark of primers4clades\ |
| table primers4clades                 | (read more                           |
| benchmark](./thumbs/primers2009.jpeg | [here](http://nar.oxfordjournals.org |
| )                                    | /cgi/content/full/gkp377v2/))        |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![phylogenetic congruence            | congruence of tree topologies        |
| heatmap](./thumbs/congruence2008.png | derived from single and concatenated |
| )                                    | loci of *Bradyrhizobium*\            |
|                                      | (read more                           |
|                                      | [here](http://www.eead.csic.es/compb |
|                                      | io/papers/Vinuesa_MLSA_Asian_Bradys_ |
|                                      | AEM08.pdf))                          |
+--------------------------------------+--------------------------------------+

</div>

<div>

+--------------------------------------+--------------------------------------+
| ![flowchart of fragment-based        | benchmark of fragment-based          |
| homology                             | comparative modelling approaches\    |
| modelling](./thumbs/fragbench_2005.p | (read more                           |
| ng)                                  | [here](http://www.eead.csic.es/compb |
|                                      | io/papers/fragbench_2005.pdf))       |
+--------------------------------------+--------------------------------------+

</div>

</div>

</div>

[](){.next}

</div>

</div>

</div>
