\

laboratory of computational & structural biology
------------------------------------------------

<div id="header">

  ----------------------------------- ------------------------------------------------------------------------ ---------------------------------------- ------------------------------------
  ![logo CSIC](./pics/logoCSIC.png)   [Estación Experimental Aula Dei](http://www.eead.csic.es/)\              [Fundación ARAID](http://www.araid.es)   ![logo ARID](./pics/logoARAID.gif)
                                      [Consejo Superior de Investigaciones Científicas](http://www.csic.es/)                                            
  ----------------------------------- ------------------------------------------------------------------------ ---------------------------------------- ------------------------------------

</div>

<div id="content">

-   [Home](index.html)
-   [Lab members](staff.html)
-   [Our collaborators](collaborators.html)
-   [Publications](publications_computational_biology_bioinformatics.html)
-   [Software](software_computational_biology_bioinformatics.html)
-   [Material
    educativo](material_didactico_biologia_computacional_bioinformatica.html)

\
\
### Peer-reviewed articles

1.  **Yruela I**, **Contreras-Moreira B**, Dunker AK and Niklas
    KJ (2018) Evolution of Protein Ductility in Duplicated Genes
    of Plants. *Front. Plant Sci.*
    [doi:10.3389/fpls.2018.01216](https://doi.org/10.3389/fpls.2018.01216)

2.  Nguyen NTT\*, **Contreras-Moreira B\***, Castro-Mondragon JA,
    Santana-Garcia W, Ossio R, Robles-Espinoza CD, Bahin M, Collombet S,
    Vincens P, Thieffry D, van Helden J, Medina-Rivera A,
    Thomas-Chollier M (2018) RSAT 2018: regulatory sequence analysis
    tools 20th anniversary. *Nucleic Acids Research*,
    [doi:10.1093/nar/gky317](https://doi.org/10.1093/nar/gky317)

3.  Casas AM, **Contreras-Moreira B, Cantalapiedra CP**, Sakuma S,
    Gracia MP, Moralejo M, Molina-Cano JL, Komatsuda T, Igartua E (2018)
    Resequencing Vrs1 gene in Spanish barley landraces revealed
    reversion of six-rowed to two-rowed spike. *Molecular Breeding*,
    [doi:10.1007/s11032-018-0816-z](https://doi.org/10.1007/s11032-018-0816-z),
    ENA sequences:
    [LT727691-LT727723](https://www.ebi.ac.uk/ena/data/view/LT727691-LT727723)

4.  Vinuesa P, Ochoa L, **Contreras-Moreira B** (2018)
    GET\_PHYLOMARKERS, a software package to select optimal orthologous
    clusters for phylogenomics and inferring pan-genome phylogenies,
    used for a critical geno-taxonomic revision of the
    genus Stenotrophomonas. *Frontiers in Microbiology*,
    [doi:10.3389/fmicb.2018.00771](https://doi.org/10.3389/fmicb.2018.00771)
    ([Docker image](https://hub.docker.com/r/csicunam/get_homologues),
    [GitHub](https://github.com/vinuesa/get_phylomarkers))

5.  Niklas KJ, Dunker AK, **Yruela I** (2018) The evolutionary origins
    of cell type diversification and the role of intrinsically
    disordered proteins. Journal of Experimental Botany, 69(7): 1437-46
    [doi:10.1093/jxb/erx493](https://doi.org/10.1093/jxb/erx493)

6.  Gordon SP, **Contreras-Moreira B**, Woods DP, 21+, Catalan P, Vogel
    JP (2017) Extensive gene content variation in the *Brachypodium
    distachyon* pan-genome correlates with population structure. *Nature
    Communications*, 2184
    [doi:10.1038/s41467-017-02292-8](https://www.nature.com/articles/s41467-017-02292-8)

7.  **Sancho R**, **Cantalapiedra CP**, López-Alvarez D, Gordon SP,
    Vogel JP, Pilar Catalán P,**Contreras-Moreira B** (2017) Comparative
    plastome genomics and phylogenomics of *Brachypodium*: flowering
    time signatures, introgression and recombination in recently
    diverged ecotypes. *New Phytologist*, doi:10.1111/nph.14926
    ([PDF](http://onlinelibrary.wiley.com/doi/10.1111/nph.14926/abstract),
    [preprint](https://www.researchgate.net/publication/321578652_Comparative_plastome_genomics_and_phylogenomics_of_Brachypodium_flowering_time_signatures_introgression_and_recombination_in_recently_diverged_ecotypes),
    ENA sequences:
    [LT222229-LT222230](https://www.ebi.ac.uk/ena/data/view/LT222229-LT222230)
    &
    [LT558582-LT558636](https://www.ebi.ac.uk/ena/data/view/LT558582-LT558636))

8.  Sein-Echaluce VC, Pallarés MC, Lostao A, **Yruela I**,
    Velázquez-Campo A, Peleato ML, Fillat MF (2017) Molecular basis for
    the integration of environmental signals by FurB from Anabaena sp.
    PCC 7120. *Biochemical Journal*, doi:10.1042/BCJ20170692
    ([PDF](http://www.biochemj.org/content/early/2017/12/04/BCJ20170692))

9.  **De Anda V\***, Zapata-Peñasco I, Poot-Hernandez AC, Eguiarte
    LE,**Contreras-Moreira B\*** and Valeria Souza V\* (2017) MEBS, a
    software platform to evaluate large (meta)genomic collections
    according to their metabolic machinery: unraveling the sulfur cycle.
    *GigaScience*, (\* corresponding author,
    [doi:10.1093/gigascience/gix096](https://doi.org/10.1093/gigascience/gix096),
    [talk](https://www.youtube.com/watch?v=q4sWE_ng-AQ))
    [![](./pics/premioValerie.png){height="18"}](pics/BioinformaticsPeerPrizeII2017-Thinkable.org.html)[![](./pics/GigaSciencePrizeTrack.png){height="38"}](pics/GigaSciencePrizeTrack.html)

10. **Cantalapiedra CP**, **Garcia-Pereira MJ**, Gracia MP, Igartua E,
    Casas AM, **Contreras-Moreira B** (2017) Large Differences in Gene
    Expression Responses to Drought and Heat Stress between Elite Barley
    Cultivar Scarlett and a Spanish Landrace. Front. Plant Sci. | doi:
    10.3389/fpls.2017.00647
    ([PDF](http://journal.frontiersin.org/article/10.3389/fpls.2017.00647/abstract),
    ENA sequences:
    [PRJEB12540](https://www.ebi.ac.uk/ena/data/view/PRJEB12540))

11. **Yruela I**, Oldfield CJ, Niklas KJ, Dunker AK (2017) Evidence for
    a strong correlation between transcription factor protein disorder
    and organismic complexity. Genome Biol Evol doi: 10.1093/gbe.evx073
    ([PDF](https://doi.org/10.1093/gbe/evx073))

12. **Contreras-Moreira B**, **Cantalapiedra CP**, **García-Pereira M**,
    Gordon SP, Vogel JP, Igartua E, Casas AM, Vinuesa P (2017) Analysis
    of plant pan-genomes and transcriptomes with GET\_HOMOLOGUES-EST, a
    clustering solution for sequences of the same species. Front.
    Plant Sci. 8:184. doi: 10.3389/fpls.2017.00184
    ([PDF](http://journal.frontiersin.org/article/10.3389/fpls.2017.00184/full))

13. **Yruela I**, **Contreras-Moreira B**, Magalhaes C, Osorio NS,
    Gonzalo-Asensio J (2016) *Mycobacterium tuberculosis* Complex
    Exhibits Lineage-Specific Variations Affecting Protein Ductility and
    Epitope Recognition. *Genome Biology and Evolution* doi:
    10.1093/gbe/evw279
    ([PDF](http://gbe.oxfordjournals.org/content/early/2017/01/05/gbe.evw279.full))

14. Bernal-Bayard P, Puerto-Galá L, **Yruela I** et al. (2017) The
    photosynthetic cytochrome c550 from the diatom *Phaeodactylum
    tricornutum*. *Photosynthesis Research*
    doi:10.1007/s11120-016-0327-x
    ([PDF](https://link.springer.com/article/10.1007/s11120-016-0327-x))

15. **Ksouri N**, Jiménez S, Wells CE, **Contreras-Moreira B\***,
    Gogorcena Y\* (2016). Transcriptional Responses in root and leaf of
    Prunus persica Under Drought Stress Using RNA Sequencing. *Front.
    Plant Sci.* 7:1715. doi: 10.3389/fpls.2016.01715 (\* corresponding
    author,
    [PDF](http://journal.frontiersin.org/article/10.3389/fpls.2016.01715/abstract),
    [transcripts](http://www.ebi.ac.uk/ena/data/view/HADJ01000001-HADJ01060423))
    [![](./pics/premioNajla.png){height="30"}](pics/CIHEAMMasterthesisprize2016.html)

16. Gutierrez-Carbonell E, Takahashi D, Luthje S, Gonzalez-Reyes JA,
    Mongrand S, **Contreras-Moreira B**, Abadia A, Uemura M, Abadia J,
    Lopez-Millan AF (2016) A shotgun proteomic approach reveals that Fe
    deficiency causes marked changes in the protein profiles of plasma
    membrane and detergent resistant microdomain preparations from Beta
    vulgaris roots. *J. Proteome Res.* doi:
    10.1021/acs.jproteome.6b00026
    ([PDF](http://pubs.acs.org/doi/abs/10.1021/acs.jproteome.6b00026))

17. Perez-Palacios R, Macias-Redondo S, Climent M, **Contreras-Moreira
    B**, Muniesa P, Schoorlemmer J. (2016) In vivo chromatin targets of
    the transcription factor Yin Yang 2 in trophoblast stem cells. PLoS
    ONE 11(5): e0154268
    ([PDF](http://journals.plos.org/plosone/article?id=10.1371%2Fjournal.pone.0154268))

18. **Cantalapiedra CP**, **Contreras-Moreira B**, Silvar C, Perovic D,
    Ordon F, Gracia MP, Igartua E, Casas A (2016) A cluster of NBS-LRR
    genes resides in a barley powdery mildew resistance QTL on 7HL.
    *Plant Genome*, doi: 10.3835/plantgenome2015.10.0101
    ([PDF](https://doi.org/10.3835/plantgenome2015.10.0101))

19. Valero-Gonzalez J, Leonhard-Melief C, Lira-Navarrete E, Jimenez-Oses
    G, Hernandez-Ruiz C, Pallares MC, **Yruela I**, Vasudevan D, Lostao
    A, Corzana F, Takeuchi H, Haltiwanger RS, Hurtado-Guerrero R (2016)
    A proactive role of water molecules in acceptor recognition by
    protein O-fucosyltransferase 2 *Nature Chemical
    Biology*, 12(4):240-6. doi:10.1038/nchembio.2019
    ([PDF](http://www.nature.com/nchembio/journal/vaop/ncurrent/full/nchembio.2019.html))

20. Medeot DB, Rivero MR, Cendoya E, **Contreras-Moreira B**, Rossi FA,
    Fischer SE, Becker A, Jofré E(2016) *Sinorhizobium meliloti* low
    molecular weight phosphotyrosine phosphatase SMc02309 modifies
    activity of the UDP-glucose pyrophosphorylase ExoN involved in
    succinoglycan biosynthesis. *Microbiology*, doi:
    10.1099/mic.0.000239
    ([PDF](http://mic.microbiologyresearch.org/content/journal/micro/10.1099/mic.0.000239))

21. Kelemen Z, **Sebastian A**, Xu W, Grain D, Salsac F, Avon A, Berger
    N, Tran J, Dubreucq B, Lurin C, Lepiniec L, **Contreras-Moreira B**,
    Dubos C (2015) Analysis of the DNA-Binding Activities of the
    Arabidopsis R2R3-MYB Transcription Factor Family by One-Hybrid
    Experiments in Yeast. *PLoS ONE* 10.1371/journal.pone.0141044
    ([PDF](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0141044))

22. Medina-Rivera A, Defrance M, Sand O, Herrmann C, Castro-Mondragon J,
    Delerce J, Spinelli L, Jaeger S, Blanchet C, Vincens P, Caron C,
    Staines DM, **Contreras-Moreira B**, Artufel M, Charbonnier L,
    Hernandez C, Thieffry D, Thomas-Chollier M and van Helden J. RSAT
    2015: Regulatory Sequence Analysis Tools. *Nucl Acids Res*,
    43(W1):W50-6
    ([PDF](https://academic.oup.com/nar/article-lookup/doi/10.1093/nar/gkv362),
    [RSAT::Plants](http://plants.rsat.eu/)).

23. **Yruela I** (2015) Plant development regulation: Overview
    and perspectives. *Journal of Plant Physiology* 182:62-78
    ([PDF](http://www.sciencedirect.com/science/article/pii/S0176161715001194))

24. Ceballos-Laita L, Gutierrez-Carbonell E, Lattanzio G, Vázquez
    S,**Contreras-Moreira B**, Abadía A, Abadía J, López-Millán AF.
    (2015) Changes in the protein profile of*Beta vulgaris* leaf
    apoplastic fluid with Fe deficiency and Fe resupply. *Frontiers in
    Plant Science* 6:145
    ([PDF](http://journal.frontiersin.org/article/10.3389/fpls.2015.00145/abstract)).

25. Igartua E, Mansour E, **Cantalapiedra CP**, **Contreras-Moreira B**,
    Gracia MP, Fuster P, Escribano J, Molina-Cano JL, Moralejo M, Ciudad
    FJ, Thomas WTB, Karsai I and Casas A (2015) Selection footprints in
    barley breeding lines detected by combining genotyping-by-sequencing
    with reference genome information. *Molecular Breeding*, 35:11
    ([PDF](http://link.springer.com/article/10.1007/s11032-015-0194-8))

26. **Cantalapiedra CP**, Boudiar R, Casas AM, Igartua E and
    **Contreras-Moreira B** (2015) BARLEYMAP: physical and genetic
    mapping of nucleotide sequences and annotation of surrounding loci
    in barley. *Molecular Breeding*, 35:13
    ([PDF](http://link.springer.com/article/10.1007/s11032-015-0253-1))

27. Dubos C, Kelemen Z, **Sebastián A**, Bulow L, Huep G, Xu W, Grain D,
    Salsac F, Brousse C, Lepiniec L, Weisshaar B, **Contreras-Moreira
    B**, Hehl R (2014) Integrating bioinformatic resources to predict
    transcription factors interacting with cis-sequences conserved in
    co-regulated genes. *BMC Genomics*, 15:317
    ([PDF](http://www.biomedcentral.com/1471-2164/15/317),
    [poster](http://cdn.f1000.com/posters/docs/282097700))

28. Ito S, Magalska A, Alcaraz-Iborra M, Lopez-Atalaya JP, Rovira V,
    **Contreras-Moreira B**, Lipinski M, Olivares R, Martinez-Hernandez
    J, Ruszczycki B, Lujan R, Geijo-Barrientos E, Wilczynski GM and
    Barco A (2014) Loss of neuronal 3D chromatin organization causes
    transcriptional and behavioural deficits related to
    serotonergic dysfunction. *Nature Communications*, 5:4450
    (doi:10.1038/ncomms545015:317,
    [PDF](http://www.nature.com/ncomms/2014/140718/ncomms5450/full/ncomms5450.html))

29. Loscos J, Igartua E, **Contreras-Moreira B**, Gracia MP and Casas
    AM (2014) HvFT1 polymorphism and effect – survey of barley germplasm
    and expression analysis *Front. Plant Sci.*, doi:
    10.3389/fpls.2014.00251
    ([PDF](http://journal.frontiersin.org/Journal/10.3389/fpls.2014.00251))

30. Sebastián,A. and Contreras-Moreira,B. (2014) footprintDB: a database
    of transcription factors with annotated cis elements and
    binding interfaces. *Bioinformatics*, 30(2):258-265
    ([PDF](http://bioinformatics.oxfordjournals.org/content/30/2/258.full.pdf+html),
    [poster](http://cdn.f1000.com/posters/docs/282097700))

31. Yruela,I. and Contreras-Moreira,B. (2013) Genetic recombination is
    associated with intrinsic disorder in plant proteomes. *BMC
    Genomics*, 14: 772
    ([PDF](http://www.biomedcentral.com/1471-2164/14/772))

32. **Contreras-Moreira B**, Vinuesa P. (2013) GET\_HOMOLOGUES, a
    versatile software package for scalable and robust microbial
    pan-genome analysis. *Appl Environ Microbiol*, 79:24 7696-7701
    ([PDF](http://aem.asm.org/content/79/24/7696.full.pdf+html))

33. Hofmann K, Silvar C, Casas AM, Herz M, Büttner B, Gracia
    MP,**Contreras-Moreira B**, Wallwork H, Igartua E, Schweizer
    G (2013) Fine mapping of the Rrs1 resistance locus against scald in
    two large populations derived from Spanish barley landraces. *Theor
    Appl Genet*, Epub ahead of print
    ([PDF](http://link.springer.com/content/pdf/10.1007%2Fs00122-013-2196-4.pdf))

34. Yruela,I. (2013) Transition metals in plant photosynthesis.
    *Metallomics.*, 5(9):1090-109.
    ([PDF](http://pubs.rsc.org/en/content/articlepdf/2013/mt/c3mt00086a))

35. Serra,T.S., Figueiredo,D.D., Cordeiro,A.M., Almeida,D.M.,
    Louren¸o,T., Abreu,I.A., **Sebastián,A,** Fernandes,L.,
    **Contreras-Moreira,B.**,Oliveira,M.M. and Saibo,N.J.M. (2013)
    OsRMC, a negative regulator of salt stress response in rice, is
    regulated by two AP2/ERF transcription factors. *Plant Mol. Biol.*,
    82 (4-5): 439-455
    ([PDF](http://link.springer.com/article/10.1007/s11103-013-0073-9))

36. Sebastián,A. and Contreras-Moreira,B. (2013) The twilight zone of
    cis element alignments. *Nucleic Acids Research*, 41(3): 1438-1449
    ([PDF](http://nar.oxfordjournals.org/content/41/3/1438.abstract.html),[TFcompare](http://floresta.eead.csic.es/tfcompare/))

37. Yruela,I. and Contreras-Moreira,B. (2012) Protein disorder in
    plants: a view from the chloroplast *BMC Plant Biology*, 12:165
    ([PDF](http://www.biomedcentral.com/1471-2229/12/165))

38. Casas,A., Dejemel,A., Ciudad,F., Yahiaoui,S., Ponce,L.,
    **Contreras-Moreira,B.**, Gracia,M.P., Lasa,J.M. and Igartua,E.
    (2011) HvFT1 (VrnH3) drives latitudinal adaptation in
    Spanish barleys. *Theoretical and Applied Genetics*,
    **122**:1293-1304
    ([PDF](http://www.springerlink.com/content/761k034518wh82r8/fulltext.pdf),
    [suppl](http://www.springerlink.com/content/761k034518wh82r8/122_2011_Article_1531_ESM.html))

39. **Yruela,I.**, Arilla-Luna,S., Medina,M. and
    **Contreras-Moreira,B.** (2010) Evolutionary divergence of
    chloroplast FAD synthetase proteins. *BMC Evolutionary Biology*,
    **10**:311 ([PDF](http://www.biomedcentral.com/1471-2148/10/311))

40. Janga,S.C. and **Contreras-Moreira,B.** (2010) Dissecting the
    expression patterns of transcription factors across conditions using
    an integrated network-based approach. *Nucleic Acids Research*,
    **38(20)**:6841-6856
    ([PDF](http://nar.oxfordjournals.org/content/38/20/6841.full.pdf+html),[suppl](http://www.eead.csic.es/compbio/suppl/network_markers/))

41. **Contreras-Moreira,B.** (2010) 3D-footprint: a database for the
    structural analysis of protein-DNA complexes. *Nucleic Acids
    Research*, **38**:D91-D97
    ([PDF](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2808867/pdf/gkp781.pdf),
    [talk](./papers/3dfootprint_jbi2010.pdf), [interface
    video](http://digital.csic.es/bitstream/10261/28935/1/interface_movie.gif))

42. **Contreras-Moreira,B.**, Sancho,J., Espinosa Angarica,V. (2010)
    Comparison of DNA binding across protein superfamilies. *Proteins*,
    **78(1)**:52-62
    ([PDF](http://www3.interscience.wiley.com/cgi-bin/fulltext/122477756/PDFSTART),[talk](./papers/3dfootprint_bifi.pdf))
    ![](./pics/logo_mem_ortiz.png){height="25"}

43. Mendoza-Vargas,A., Olvera,L., Olvera,M., Grande,R.,
    Vega-Alvarado,L., Taboada,B., Jimenez-Jacinto,V., Salgado,H.,
    Juárez,K., **Contreras-Moreira,B.**, Huerta,A.M.,
    Collado-Vides,J., Morett,E. (2009) Genome-Wide Identification of
    Transcription Start Sites, Promoters and Transcription Factor
    Binding Sites in E. coli. *PLoS ONE*, **4(10)**: e7526.
    doi:10.1371/journal.pone.0007526
    ([PDF](http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0007526))

44. **Contreras-Moreira,B.**, Sachman-Ruiz,B., Figueroa-Palacios,I.,
    and Vinuesa,P. (2009) primers4clades: a web server that uses
    phylogenetic trees to design lineage-specific PCR primers for
    metagenomic and diversity studies. *Nucleic Acids Research*,
    **37**(Web Server issue):W95-W100
    ([PDF](http://nar.oxfordjournals.org/cgi/reprint/37/suppl_2/W95),
    PlantGEM [poster](./papers/PlantGem2009.pdf))

45. Espinosa Angarica,V., González Pérez,A.,
    Vasconcelos,A.T., Collado-Vides,J. and
    **Contreras-Moreira,B.** (2008) Prediction of TF target sites based
    on atomistic models of protein-DNA complexes. *BMC Bioinformatics*,
    **9**:436.
    ([PDF](http://www.biomedcentral.com/1471-2105/9/436),[talk](./papers/talk_jnb08.pdf),[PWMs](./papers/dnaprot_experimental_consensi.txt))

46. Vinuesa,P., Rojas-Jiménez,K., **Contreras-Moreira,B.**, Mahna,S.K.,
    Prasad,B.N., Moe,H., Selvaraju,S.B., Thierfelder,H., and Werner,D.
    (2008) Multilocus Sequence Analysis for assessment of the
    biogeography and evolutionary genetics of four Bradyrhizobium
    species that nodulate soybeans on the asiatic continent.
    *Appl. Environ. Microbiol.*, **74**:6987-6996
    ([PDF](./papers/Vinuesa_MLSA_Asian_Bradys_AEM08.pdf))

47. Lozada-Chávez,I., Espinosa Angarica,V., Collado-Vides,J. and
    **Contreras-Moreira,B.** (2008) The role of DNA-binding specificity
    in the evolution of bacterial regulatory networks. *Journal of
    Molecular Biology*, **379(3)**:627-43.
    ([PDF](./papers/prok_specificity.pdf),
    [suppl](./papers/prok_specificity_suppl.pdf), ISSOL2008
    [poster](./papers/prok_specificity_poster.pdf) &
    [abstract](http://www.springerlink.com/content/q3110467090237v5/?p=1b4e1f52cb4f44e098d12fefd8596631&pi=0))

48. Gama-Castro,S., Jiménez-Jacinto,V., Peralta-Gil,M.,
    Santos-Zavaleta,A., Peñaloza-Spinola,M.I., **Contreras-Moreira,B.**
    \[..\] & Collado-Vides,J. (2008) RegulonDB (version 6.0): Gene
    regulation model of Escherichia coli K-12 beyond transcription,
    active (experimental) annotated promoters and Textpresso navigation.
    *Nucleic Acids Research*, **36**: D120-D124
    ([PDF](./papers/regulondb2008.pdf))

49. Han, R., Leo-Macias, A., Zerbino, D., Bastolla, U.,
    **Contreras-Moreira, B.** and Ortiz, A.R.(2007) An efficient
    conformational sampling method for homology modelling. *Proteins*,
    **71(1)**: 175-188 ([PDF](./papers/CMsampling2007.pdf))

50. **Contreras-Moreira, B.**, Branger, P.A. and Collado-Vides, J.
    (2007) TFmodeller: comparative modelling of protein-DNA complexes.
    *Bioinformatics*, **23(13)**: 1694-1696
    ([PDF](./papers/tfmodeller2007.pdf),
    [suppl](http://maya.ccg.unam.mx/~tfmodell/tutorial.html),
    [poster](./papers/tfmodeller_PlantGem2007.pdf))

51. **Contreras-Moreira, B.** and Collado-Vides, J. (2006) Comparative
    footprinting of DNA-binding proteins. *Bioinformatics*, **22(14)**:
    e74.([PDF](http://bioinformatics.oxfordjournals.org/cgi/reprint/22/14/e74?ijkey=4779A1VH6G1MyzB&keytype=ref),
    [talk](./papers/dnasite_ismb2006.pdf),
    [poster](./papers/tfmodeller_NBCR2007.pdf),
    [suppl](./suppl/ismb2006/))
    ![](./pics/logo_ismb_fortaleza.png){height="25"}

52. **Contreras-Moreira, B.**, Ezkurdia, I., Tress, M.L. and Valencia,A.
    (2005) Empirical limits for template-based protein structure
    prediction: the CASP5 example. *FEBS Letters*, **579(5)**:
    1203-1207.([PDF](./papers/fragbench_2005.pdf))

53. Smith, G.R.\*, **Contreras-Moreira, B.\***, Zhang, X. and
    Bates, P.A. (2004) A link between sequence conservation and domain
    motion within the AAA+ family. *Journal of Structural Biology*,
    **146(1-2)**: 189-204 (\* joint first authors,
    [PDF](./papers/AAApaper2004.pdf)) (\* joint first authors)

54. **Contreras-Moreira, B.**, Fitzjohn, P.W., Offman, M., Smith, G.R.
    and Bates, P.A. (2003) Novel use of a genetic algorithm for protein
    structure prediction: searching template and sequence
    alignment space. *Proteins*, **53 (S6)**: 424-429.
    ([PDF](./papers/CASP5_paper2003.pdf),
    [talk](./papers/sac-casp5.pdf))
    ![](./pics/logo_casp5.gif){height="25"}

55. **Contreras-Moreira, B.\***, Jonsson, P.F.\* and Bates, P.A. (2003).
    Structural context to exons in protein domains: implications for
    protein modelling and design. *Journal of Molecular
    Biology*,**333**: 1057-1071. ([PDF](./papers/exons_paper2003.pdf))
    (\* joint first authors)

56. Jimenez, J.L., Smith, G.R., **Contreras-Moreira, B.**, Sgouros,
    J.G.,Meunier, F.A., Bates P.A. and Schiavo, G. (2003). A comparative
    study of synaptotagmin, protein kinase C and phosholipase C by
    sequence, structural and modelling approaches. *Journal of Molecular
    Biology*,**333**: 621-639.
    ([PDF](http://www.sciencedirect.com/science/article/pii/S0022283603010957))

57. Sellar,G.C., Watt,K.P., Rabiasz,G.J., Stronach,E.A., Li,L.,
    Miller,E.P.,Massie,C.E., Miller,J., **Contreras-Moreira,B.**,
    Scott,D., Brown,I., Williams,A.R., Bates,P.A., Smyth,J.F.
    and Gabra,H. (2003). OPCML at 11q25 is epigenetically inactivated
    and has tumor-suppressor function in epithelial ovarian cancer.
    *Nature Genetics*, **34**: 337-343.
    ([PDF](http://www.nature.com/ng/journal/v34/n3/full/ng1183.html))

58. **Contreras-Moreira,B.**, Fitzjohn,P.W. and Bates,P.A. (2003). *In
    silico* Protein Recombination: enhancing template and sequence
    alignment selection for comparative protein modelling. *Journal of
    Molecular Biology*,**328**:
    593-608.([PDF](./papers/recomb2003.pdf),[talk](./papers/corunna2003.pdf))

59. **Contreras-Moreira,B.**, Fitzjohn,P.W. and Bates,P.A. (2002).
    Comparative modelling: an essential methodology for protein
    structure prediction in the post-genomic era. *Applied
    Bioinformatics*,**1(4)**;177-190.([PDF](./papers/review_CM2002.pdf))

60. **Contreras-Moreira,B.** and Bates,P.A. (2002). Domain Fishing: a
    first step in protein comparative modelling. *Bioinformatics*
    **18**: 1141-1142.
    ([PDF](./papers/domfish2002.pdf),[talk](./papers/salamanca2002.pdf))

### Books and chapters

i.  Contreras-Moreira B (2018) Algoritmos en Bioinformática Estructural.
    doi:[10.20350/digitalcsic/8544](https://doi.org/10.20350/digitalcsic/8544).

ii. **Contreras-Moreira B**, Castro-Mondragon J, Rioualen C,
    **Cantalapiedra CP**, van Helden J (2016) RSAT::Plants: Motif
    Discovery within Clusters of Upstream Sequences in Plant Genomes. In
    Plant Synthetic Promoters: Methods and Protocols, edited by Hehl R.
    *Methods in Molecular Biology*, 1482:279-95
    ([PDF](https://www.ncbi.nlm.nih.gov/pubmed/27557774), [author's
    PDF](https://digital.csic.es/handle/10261/136253))

iii. Castro-Mondragon J, Rioualen C, **Contreras-Moreira B**, van Helden
    J (2016) RSAT::Plants: Motif Discovery in ChIP-Seq Peaks of
    Plant Genomes. In Plant Synthetic Promoters: Methods and Protocols,
    edited by Hehl R. *Methods in Molecular Biology*, 1482:297-322
    ([PDF](https://www.ncbi.nlm.nih.gov/pubmed/27557775), [author's
    PDF](https://digital.csic.es/handle/10261/136254))

iv. **Contreras-Moreira B** and **Sebastian A** (2016) footprintDB:
    analysis of plant cis regulatory elements, transcription factors and
    binding interfaces. In Plant Synthetic Promoters: Methods and
    Protocols, edited by Hehl R. *Methods in Molecular Biology*,
    1482:259-77
    ([PDF](http://link.springer.com/protocol/10.1007%2F978-1-4939-6396-6_17),
    [author's PDF](https://digital.csic.es/handle/10261/136262))

v.  **Yruela I**(2016) Las proteínas dúctiles. Editorial CSIC
    y Catarata. ISBN: 978-84-00-10055-1. Pages: 94
    ([comprar](http://editorial.csic.es/publicaciones/libros/12685/978-84-00-10055-1/las-proteinas-ductiles.html))

vi. Vinuesa P and **Contreras-Moreira B** (2015) Robust Identification
    of Orthologues and Paralogues for Microbial Pan-Genomics Using
    GET\_HOMOLOGUES: A Case Study of pIncA/C Plasmids. In Bacterial
    Pangenomics, edited by A Mengoni, M Galardini and M Fondi. Methods
    in Molecular Biology, 1231: 203-232.
    ([PDF](http://link.springer.com/protocol/10.1007/978-1-4939-1720-4_14))

vii. Vinuesa P, **Contreras-Moreira B** (2015) Chapter 27: Pangenomic
    analysis of the Rhizobiales using the GET\_HOMOLOGUES
    software package. In Biological Nitrogen Fixation 7 (I), p271-80,
    Frans J de Bruijn (ed). Wiley.
    ([PDF](https://doi.org/10.1002/9781119053095.ch27))

viii. **Yruela I** (2014) Proteínas desordenadas. In Bioinformática con
    Ñ, edited by Sebastián A. and Pascual-García A. Zenodo
    ([doi:10.5281/zenodo.1066352](http://doi.org/10.5281/zenodo.1066352))

ix. **Yruela I**, **Sebastián A** (2014) Macromoléculas biológicas:
    proteínas, DNA y RNA. In Bioinformática con Ñ, edited by
    Sebastián A. and Pascual-García A. Zenodo
    ([doi:10.5281/zenodo.1067867](http://doi.org/10.5281/zenodo.1067867))

x.  **Sebastián A** (2014) Análisis de secuencias biológicas. In
    Bioinformática con Ñ, edited by Sebastián A. and Pascual-García A.
    Zenodo
    ([doi:10.5281/zenodo.1065598](http://doi.org/10.5281/zenodo.1065598))

xi. **Sebastián A**, Fernández JM (2014) Computación paralela y
    clústeres de cálculo. In Bioinformática con Ñ, edited by
    Sebastián A. and Pascual-García A. Zenodo
    ([doi:10.5281/zenodo.1065592](http://doi.org/10.5281/zenodo.1065592))

xii. Fernández JM, **Sebastián A** (2014) Gestión de bibliografía e
    información y escritura de artículos científicos. In Bioinformática
    con Ñ, edited by Sebastián A. and Pascual-García A. Zenodo
    ([doi:10.5281/zenodo.1065035](http://doi.org/10.5281/zenodo.1065035))

xiii. Sebastián,A., Cantalapiedra,C.P. and Contreras-Moreira,B. (2012)
    Interface similarity improves comparison of DNA-binding proteins:
    the Homeobox example. *Lecture Notes in Computer Science*,
    6620/2012:72-82
    ([PDF](http://www.springerlink.com/content/p21pk20883843p04/),
    [authors' PDF](./papers/LNCS_2012_6620_72-82.pdf),
    [suppl](http://www.eead.csic.es/compbio/suppl/homeobox_interface/data.xls))
    ![](./pics/logo_jbi2010.png){height="25"}

xiv. Sachman-Ruiz,B. **Contreras-Moreira,B.**,
    Zozaya,E., Martínez-Garza,C. and Vinuesa,P. (2011) primers4clades, a
    web server to design lineage-specfic PCR primers for
    gene-targeted metagenomics. **Chapter 51** in Frans J. de Bruijn
    (ed.), *Handbook of Molecular Microbial Ecology vol.I: Metagenomics
    and Complementary Approaches*, ([Google
    books](http://books.google.es/books?id=v9jzjED3QJoC&lpg=PA444&dq=primers4clades&pg=PA441#v=onepage&q=primers4clades&f=false),[Wiley](http://eu.wiley.com/WileyCDA/WileyTitle/productCd-0470644796,descCd-tableOfContents.html))

xv. Abascal F, Alonso-Allende R, Blaschke C, **Contreras-Moreira B**
    \[...\] and Valencia A (2001) Bioinformatics For Functional and
    Structural Genomics at the Protein Design group, CNB-CSIC. In Rita
    Casadio, Lanfranco Masotti (ed) Protein Sequence Analysis In The
    Post-Genomic Era: 149-165. CLUEB.

### PhD & Master theses

-   **Anda Torres, VY** (2018) La metagenómica como herramienta de
    conservación de Cuatro Ciénegas Coahuila: el caso del ciclo
    del azufre. PhD thesis, Universidad Nacional Autónoma de México
    ([PDF](http://oreon.dgbiblio.unam.mx/F/L1S8MKDS5TTA4TR11NBHPPD6JE83RLANXRE5RRSMBGGTQP4UYY-09317?func=full-set-set&set_number=001392&set_entry=000001&format=999))
-   **Cantalapiedra CP** (2016) Accessing genetic variability in Spanish
    barleys through high-throughput sequencing. PhD thesis, Universitat
    Autonoma de Barcelona
    ([PDF](https://digital.csic.es/handle/10261/141791))
-   **Ksouri N** (2016) Transcriptional Responses in Prunus persica
    under Conditions of Drought Stress using RNA Sequencing. Master
    thesis, CIHEAM-Universitat de Lleida
    ([PDF](http://www.ciheam.org/uploads/attachments/572/3._KSOURI_Full_Thesis.pdf),
    [talk](http://www.ciheam.org/uploads/attachments/571/4._KSOURI_Thesis_PPT_presentation.pdf))
-   **Contreras-Moreira B** (2004). Algorithms for protein comparative
    modelling and some evolutionary implications. PhD thesis
    (Biochemistry), University of London.
    ([PDF](./papers/PhD_thesis_Bruno_Contreras-Moreira2003.pdf),
    [talk](./papers/exit_cruk2004.pdf))

</div>
